tools/issue-labeler/backfill_test.go

```diff
@@ -0,0 +1,168 @@
+package main
+
+import (
+	"fmt"
+
+	"reflect"
+	"regexp"
+	"strings"
+	"testing"
+)
+
+func testIssueBodyWithResources(resources []string) string {
+	return fmt.Sprintf(`
+### New or Affected Resource(s):
+
+%s
+
+#
+`, strings.Join(resources, "\n"))
+}
+
+func TestComputeIssueUpdates(t *testing.T) {
+	defaultRegexpLabels := []regexpLabel{
+		{
+			Regexp: regexp.MustCompile("google_service1_.*"),
+			Label:  "service/service1",
+		},
+		{
+			Regexp: regexp.MustCompile("google_service2_resource1"),
+			Label:  "service/service2-subteam1",
+		},
+		{
+			Regexp: regexp.MustCompile("google_service2_resource2"),
+			Label:  "service/service2-subteam2",
+		},
+	}
+	cases := map[string]struct {
+		issues               []Issue
+		regexpLabels         []regexpLabel
+		expectedIssueUpdates []IssueUpdate
+	}{
+		"no issues -> no updates": {
+			issues:               []Issue{},
+			regexpLabels:         defaultRegexpLabels,
+			expectedIssueUpdates: []IssueUpdate{},
+		},
+		"exempt labels -> no updates": {
+			issues: []Issue{
+				{
+					Number:      1,
+					Body:        testIssueBodyWithResources([]string{"google_service1_resource1"}),
+					Labels:      []Label{{Name: "service/terraform"}},
+					PullRequest: map[string]any{},
+				},
+				{
+					Number:      2,
+					Body:        testIssueBodyWithResources([]string{"google_service1_resource1"}),
+					Labels:      []Label{{Name: "forward/exempt"}},
+					PullRequest: map[string]any{},
+				},
+			},
+			regexpLabels:         defaultRegexpLabels,
+			expectedIssueUpdates: []IssueUpdate{},
+		},
+		"add resource & review labels": {
+			issues: []Issue{
+				{
+					Number:      1,
+					Body:        testIssueBodyWithResources([]string{"google_service1_resource1"}),
+					Labels:      []Label{},
+					PullRequest: map[string]any{},
+				},
+				{
+					Number:      2,
+					Body:        testIssueBodyWithResources([]string{"google_service2_resource1"}),
+					Labels:      []Label{},
+					PullRequest: map[string]any{},
+				},
+			},
+			regexpLabels: defaultRegexpLabels,
+			expectedIssueUpdates: []IssueUpdate{
+				{
+					Number: 1,
+					Labels: []string{"forward/review", "service/service1"},
+				},
+				{
+					Number: 2,
+					Labels: []string{"forward/review", "service/service2-subteam1"},
+				},
+			},
+		},
+		// We may need to change this behavior later in order to ensure these tickets get forwarded.
```

Comment:
@melinath in case you see this note post-merge, I think this behavior is correct and should not need to change. However, we will need to spot check that the service labels on older tickets are correct, because they will not be removed and are assumed to be properly assigned (currently 154 such tickets).

Reply:
@roaks3 just to make sure we're talking about the same thing, my concern is that old tickets that are already correctly labeled will never be forwarded during backfill.

Reply:
Yea, that was my understanding of your comment. To clarify the current design: the presence of a valid service label will enable old tickets to be forwarded (unless there are other labels preventing forwarding), but currently we are not running our internal job against tickets before the labeler was introduced. Once we start the backfill process, we will be running the labeler and internal job on the same set of tickets, and they would be forwarded even if the labeler did not make any change.